### PR TITLE
Feature/ausclanwar part1 test and function

### LINKED ISSUE #47: !ausclanwar command to find which players have/haven't been active in war in past 4 weeks
Can use the following endpoint to determine player participation over recent 4 weeks.
```
GET /clans/{clanTag}/riverracelog
```
Note: endpoint is from official Royale API https://developer.clashroyale.com/ and seems to only return data for completed war week instead of current war week (double check this)

### Criteria for "actively participating in war":
1. Been in clan for 4 weeks
2. Have done at least 3 of 4 war days per week (i.e. used 4 decks 3 times = 12 battles. At a minimum if they lose every battle that's NOT a boat attack, that would be 12 * 100 = 1200 fame. Use `fame` instead of `decksUsed` since can only get `fame` on war days, whereas `decksUsed` include practise days. ~~[TBD - check slack chat for final decision] Have done at least 3 of 4 war days per week (i.e. used 4 decks 3 times = 12 `decksUsed` that week)~~

Note: Discuss with @damoster , we might be able to make this limit dynamic. Looks like API response shows season start time, and what time a clan finishes the race. So we could calc how many war days it took for us to win (or lose) and set a required battle limit accordingly.... (Though maybe lets make that a separate issue/PR and try to keep this one as simple/quick as possible)

### Output requirements
- "X" if there was any week which had fame less than 1200. "/" if all weeks were above minimum fame.
- Can call `GET /clans/{clanTag}/members` to get the roles
- Order output by Ave fame per week, have the following columns (Or maybe order by min Fame in any of past 4 Weeks of each player)
```
WarActive | PlayerName | Role     |  Past4WeekFameCount  |
----------+------------+----------+----------------------+
    /     | SomeDude   | Member   |  2700,2400,2500,2600 |
    /     | D-drizzle  | Co-leader|  1600,1500,2000,1900 |
    X     | JoeMama    | Elder    |  2000,1400, 900,1000 |
    X     | JoeDad     | Elder    |  1000, 200,1400,1800 |
```
- Make sure spacing is nice for the `Past4WeekFameCount`, i.e. even if number is 3 digits, make it expand to take up 4 spaces. There's examples online and how @damoster has done it for existing command outputs (have a look)
- I know there's a size limit on output for urls, so maybe for now, just use the letters "/" and "X". Can look into icons later in a separate issue/PR, I saw deckshoppro.com bot doing this
image

If showing entire clan exceeds message limit, then we should ONLY show: 
1. Elder's who HAVE NOT been "actively participating in war"
2. Member's who HAVE been "actively participating in war"

### How to split up work
Split this work into 3 separate sub-issues or 3 separate PRs:
1. Tests + Function/class to process output from `GET /clans/{clanTag}/riverracelog` into a data structure ready for final output requirements. Btw, I know we've been using dictionaries the whole time, but can you use a list of dataclasses instead? https://docs.python.org/3/library/dataclasses.html
2. API call to `GET /clans/{clanTag}/riverracelog`
3. Add discord bot command. Lets call it `!ausclanwar` for ausclan specific, and `!waractive [clantag]` for any clan

### PR DESCRIPTION
**Pull request by josephwccheng**
closes #47

### CODE REVIEW

File: clash_royale_client.py [hunk 1]
```diff
@@ -33,3 +33,7 @@ def get_player_info(self, player_tag):
         cleaned_tag = player_tag.replace('#', '').upper()
         request_url = '{}/players/%23{}'.format(self.base_url, cleaned_tag)
         return self.make_request(request_url, 'get_player_info')
+
+    # url param pre-conditoin -> clan_tag should not have # and all caps
+    def get_river_race_log(self, clan_tag):
+        return True
```
Comment (damoster) [inline clash_royale_client.py:39]: Doesn't really matter but for empty functions yet to be implemented, you could do one of these instead
```
def get_river_race_log(self, clan_tag):
    # TODO
    pass
```
or
```
def get_river_race_log(self, clan_tag):
    # TODO
    return
```

File: common/schemas.py [hunk 2]
```diff
@@ -1,4 +1,17 @@
+from dataclasses import dataclass
+
 # Numbers represent the prefix of the id of the cards.
 # i.e. 26000037 is a Inferno Dragon, 27000007 is Elixir Collector, 28000000 is fireball
 CARD_TYPE_ID_PREFIX = {'26': 'troop', '27': 'building', '28': 'spell'}
 MAX_CARD_LEVEL = 14
+
+# Data Clases
+@dataclass
+class player_historical_activity:
```
Comment (damoster) [inline common/schemas.py:10]: Python class names should be camel case instead of snake case. Please fix

File: clash_royale_service.py [hunk 3]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
         member_cards_ranked = self.sort_list_by_card_level(member_cards_ranked)
 
         return clan_info, member_cards_ranked
+    '''
+    River Race Player Activity History Functions
+    '''
+    def get_river_race_log(self, clan_tag: str):
```
Comment (damoster) [inline clash_royale_service.py:108]: What's the point of this function? I don't see it being used anywhere

File: clash_royale_service.py [hunk 4]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
         member_cards_ranked = self.sort_list_by_card_level(member_cards_ranked)
 
         return clan_info, member_cards_ranked
+    '''
+    River Race Player Activity History Functions
+    '''
+    def get_river_race_log(self, clan_tag: str):
+        return self.clash_royale_client.get_river_race_log(clan_tag)
+
+    def clan_river_race_history(self, clan_tag: str, past_weeks=4) -> Dict[str, player_historical_activity]:
```
Comment (damoster) [inline clash_royale_service.py:111]: /nit inconsistent spacing for `past_weeks=4`, whereas function below you have `past_weeks = 4`

Also, missing type hints, so should be `past_weeks: int = 4`

File: clash_royale_service.py [hunk 5]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
...
+                boat_attacks_hist = [None]* past_weeks,
+                deck_used_hist = [None]* past_weeks
+            )
+        return self.past_weeks_clan_war(clan_players_war_history, clan_tag, past_weeks)
+
+    def past_weeks_clan_war(self, clan_players_war_history: Dict[str, player_historical_activity], clan_tag: str, past_weeks = 4) -> Dict[str, player_historical_activity]:
+        client_race_log_api_response = self.clash_royale_client.get_river_race_log(clan_tag)
+        grouped_clan_wars = [item for index, item in enumerate(client_race_log_api_response['items']) if index < past_weeks]
+        current_week = 0
+        #First loop splits the clan wars into specific week
```
Comment (damoster) [inline clash_royale_service.py:131]: /supernit space after #

File: clash_royale_service.py [hunk 6]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
...
+                exp_level = member['expLevel'],
+                fame_hist = [None]* past_weeks,
+                boat_attacks_hist = [None]* past_weeks,
+                deck_used_hist = [None]* past_weeks
+            )
+        return self.past_weeks_clan_war(clan_players_war_history, clan_tag, past_weeks)
+
+    def past_weeks_clan_war(self, clan_players_war_history: Dict[str, player_historical_activity], clan_tag: str, past_weeks = 4) -> Dict[str, player_historical_activity]:
+        client_race_log_api_response = self.clash_royale_client.get_river_race_log(clan_tag)
+        grouped_clan_wars = [item for index, item in enumerate(client_race_log_api_response['items']) if index < past_weeks]
```
Comment (damoster) [inline clash_royale_service.py:129]: Would it be simpler to just do this instead using python list slicing?
```
client_race_log_api_response['items'][:past_weeks]
```

File: clash_royale_service.py [hunk 7]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
...
+                clan = standing['clan']
+                if clan['tag'] == clan_tag:
+                    # Third Grab all the participants from the list and store it into the output object
+                    for participant in clan['participants']:
+                        if participant['tag'] in clan_players_war_history:
+                            clan_players_war_history[participant['tag']].fame_hist[current_week] = participant['fame']
+                            clan_players_war_history[participant['tag']].boat_attacks_hist[current_week] = participant['boatAttacks']
+                            clan_players_war_history[participant['tag']].deck_used_hist[current_week] = participant['decksUsed']
+                    break
+            current_week = current_week + 1
```
Comment (damoster) [inline clash_royale_service.py:144]: Fyi, while you can't do `current_week++` in python, you can do this for shorter notation: `current_week += 1`

File: tests/test_clan_members_war.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+        clan_tag = "#9GULPJ9L"
+        # Mock API responses and certain function calls
+        when(clan_members_war.clash_royale_client).get_river_race_log(clan_tag).thenReturn(
+            clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE
+        )
+
+        clan_players_war_history = clan_members_war.past_weeks_clan_war(mock_clan_players_war_history, clan_tag)
+        # Testing for the length returned. Should be 3 players returned
+        self.assertEqual(len(clan_players_war_history), 4)
+        pass
```
Comment (damoster) [inline tests/test_clan_members_war.py:28]: lmao, your test never reaches the below asserts since you have this `pass` here. Remove unnecessary pass

File: clash_royale_service.py [hunk 9]
```diff
@@ -33,12 +34,16 @@ def compare_card_levels(p1, p2):
     return compare_result
 
 
-class ClanMembersRanker:
+class ClashRoyaleService:
```
Comment (damoster) [inline clash_royale_service.py:37]: You also need to do this renaming in the discord commands or main .py files too don't you? Otherwise existing commands would fail

File: common/schemas.py [hunk 10]
```diff
@@ -1,4 +1,17 @@
+from dataclasses import dataclass
+
 # Numbers represent the prefix of the id of the cards.
 # i.e. 26000037 is a Inferno Dragon, 27000007 is Elixir Collector, 28000000 is fireball
 CARD_TYPE_ID_PREFIX = {'26': 'troop', '27': 'building', '28': 'spell'}
 MAX_CARD_LEVEL = 14
+
+# Data Clases
```
Comment (damoster) [inline common/schemas.py:8]: Is this comment necessary? hahahaha The annotation already indicates that it's a dataclass lmaooo

File: tests/test_clan_members_war.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+        clan_players_war_history = clan_members_war.clan_river_race_history(clan_tag)
+        self.assertEqual(len(clan_players_war_history), 3)
+        self.assertEqual(clan_players_war_history['#8VUG0GQRY'].fame_hist, [2650, 2800, 2500, 3100])
+        self.assertEqual(clan_players_war_history['#LYJVYUUUR'].fame_hist, [2800, 2200, 2250, 3100])
+
+        # Testing for changing the past week to longer period i.e. 8
+        clan_players_war_history = clan_members_war.clan_river_race_history(clan_tag, 8)
+        self.assertEqual(len(clan_players_war_history['#8VUG0GQRY'].fame_hist), 8)
+        pass
```
Comment (damoster) [inline tests/test_clan_members_war.py:55]: remove redundant `pass`

File: tests/test_clan_members_war.py [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    def test_past_weeks_clan_war(self):
+        clan_members_war = ClashRoyaleService()
+        clan_tag = "#9GULPJ9L"
+        # Mock API responses and certain function calls
+        when(clan_members_war.clash_royale_client).get_river_race_log(clan_tag).thenReturn(
+            clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE
+        )
+
+        clan_players_war_history = clan_members_war.past_weeks_clan_war(mock_clan_players_war_history, clan_tag)
+        # Testing for the length returned. Should be 3 players returned
```
Comment (damoster) [inline tests/test_clan_members_war.py:26]: You mean 4 in this case? XD

File: tests/test_clan_members_war.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+        self.assertEqual(len(clan_players_war_history), 4)
+        pass
+
+        # Testing if fame_hist is updated for Joseph, Zepol, and Someone Who hasnt done War
+        self.assertEqual(clan_players_war_history['#8VUG0GQRY'].fame_hist, [2650, 2800, 2500, 3100])
+        self.assertEqual(clan_players_war_history['#LYJVYUUUR'].fame_hist, [2800, 2200, 2250, 3100])
+        self.assertEqual(clan_players_war_history['#SOMENOBDY'].fame_hist, [None, None, None, None])
```
Comment (damoster) [inline tests/test_clan_members_war.py:33]: Can you also add a case where some only does some of the wars? You could either:
1. Find someone who joined recently so only did like say past 2 of 4 wars
2. Edit the output of `clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE` and make someone who did all wars appear to only have done 2 of 4 wars or something

File: tests/test_clan_members_war.py [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+            clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE
+        )
+
+        clan_players_war_history = clan_members_war.past_weeks_clan_war(mock_clan_players_war_history, clan_tag)
+        # Testing for the length returned. Should be 3 players returned
+        self.assertEqual(len(clan_players_war_history), 4)
+        pass
+
+        # Testing if fame_hist is updated for Joseph, Zepol, and Someone Who hasnt done War
+        self.assertEqual(clan_players_war_history['#8VUG0GQRY'].fame_hist, [2650, 2800, 2500, 3100])
```
Comment (damoster) [inline tests/test_clan_members_war.py:31]: What about checking and comparing with the other fields? i.e. `boat_attacks_hist`, `deck_used_hist` etc

Comment (damoster) [reply tests/test_clan_members_war.py:31]: e.g. you could do something like this
```
expected_player_goku_history = player_historical_activity(tag='#YV9GU2VG', name='Goku', role='member', exp_level=13, fame_hist=[2650, 2800, 2500, 3100], boat_attacks_hist=[None, None, None, None], deck_used_hist=[16, 18, 18, 12])
self.assertEqual(clan_players_war_history['#8VUG0GQRY'], expected_player_goku_history)
```

File: tests/test_clan_members_war.py [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+            clash_royale_client_responses.CLAN_INFO_API_RESPONSE
+        )
+        when(clan_members_war.clash_royale_client).get_river_race_log(clan_tag).thenReturn(
+            clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE
+        )
+
+        clan_players_war_history = clan_members_war.clan_river_race_history(clan_tag)
+        self.assertEqual(len(clan_players_war_history), 3)
+        self.assertEqual(clan_players_war_history['#8VUG0GQRY'].fame_hist, [2650, 2800, 2500, 3100])
+        self.assertEqual(clan_players_war_history['#LYJVYUUUR'].fame_hist, [2800, 2200, 2250, 3100])
```
Comment (damoster) [inline tests/test_clan_members_war.py:50]: Same deal here, see comment here https://github.com/damoster/royale_clan_card_level_ranker_bot/pull/49/files#r792305241

File: clash_royale_service.py [hunk 16]
```diff
@@ -97,3 +102,44 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
...
+                if clan['tag'] == clan_tag:
+                    # Third Grab all the participants from the list and store it into the output object
+                    for participant in clan['participants']:
+                        if participant['tag'] in clan_players_war_history:
+                            clan_players_war_history[participant['tag']].fame_hist[current_week] = participant['fame']
+                            clan_players_war_history[participant['tag']].boat_attacks_hist[current_week] = participant['boatAttacks']
+                            clan_players_war_history[participant['tag']].deck_used_hist[current_week] = participant['decksUsed']
+                    break
+            current_week = current_week + 1
+        return clan_players_war_history
```
Comment (damoster) [inline clash_royale_service.py:145]: Before we return, can then loop through the updated clan_players_war_history and compute the `WarActive` value (add it to the dataclass). Makes sense to do it here rather than in the discord command level

Also, please add tests to compare/check this for:
- Case Active
- Case Not active

File: common/schemas.py [hunk 17]
```diff
@@ -1,4 +1,17 @@
...
+@dataclass
+class player_historical_activity:
+    tag: str
+    name: str
+    role: str
+    exp_level: int
+    fame_hist: list
+    boat_attacks_hist: list
```
Comment (damoster) [inline common/schemas.py:16]: What's this for? Same with `deck_used_hist`. Aren't those fields out of scope for the requirements outlined in the ticket?

**Review by damoster — CHANGES_REQUESTED** [review]
Looking good. A few fixes needed before merge ready tho

File: common/schemas.py [hunk 18]
```diff
@@ -1,4 +1,46 @@
...
+
 # Numbers represent the prefix of the id of the cards.
 # i.e. 26000037 is a Inferno Dragon, 27000007 is Elixir Collector, 28000000 is fireball
 CARD_TYPE_ID_PREFIX = {'26': 'troop', '27': 'building', '28': 'spell'}
 MAX_CARD_LEVEL = 14
+
+@dataclass
+class PlayerActivity:
+    '''
+    A player has a  "WarActiveWeek" for a specific week IF:n
```
Comment (damoster) [inline common/schemas.py:11]: /supernit I know I wrote this lol. But pls fix, should be `if:` instead of `IF:n`

File: tests/test_clan_members_war.py (hunk 13 above)
Comment (josephwccheng) [reply tests/test_clan_members_war.py]: done

File: tests/test_clan_members_war.py (hunk 14 above)
Comment (josephwccheng) [reply tests/test_clan_members_war.py]: done

File: clash_royale_service.py (hunk 16 above)
Comment (josephwccheng) [reply clash_royale_service.py:175]: done - war active, elder worthy, and average fame

**Review by josephwccheng — COMMENTED** [review]
completed all comments

File: clash_royale_service.py [hunk 19]
```diff
@@ -33,12 +34,43 @@ def compare_card_levels(p1, p2):
...
+    for week in fame_hist:
+        if week is None or week < 1200:
+            elder_worthy = False
+            break
+    return elder_worthy
+
+def compute_average_fame(fame_hist: list) -> int:
+    average_fame = -1
+    result = []
+    for fame in fame_hist:
```
Comment (damoster) [inline clash_royale_service.py:58]: /nit could replace this loop with a python one liner
```python
result = [fame for fame in fame_hist if fame is not None]
```

File: clash_royale_service.py [hunk 20]
```diff
@@ -33,12 +34,43 @@ def compare_card_levels(p1, p2):
     return compare_result
 
 
-class ClanMembersRanker:
+def compute_war_active(fame_hist: list) -> bool:
+    war_active = True
+    for week in fame_hist:
+        if week is None:
+            continue
+        elif week < 1200:
```
Comment (damoster) [inline clash_royale_service.py:42]: Plz put this 1200 threshold in the schema.py and copy paste the comment explaining where this value comes from

File: clash_royale_service.py [hunk 21]
```diff
@@ -33,12 +34,43 @@ def compare_card_levels(p1, p2):
...
+            continue
+        elif week < 1200:
+            war_active = False
+            break
+    return war_active
+
+def compute_elder_worthy(fame_hist: list) -> bool:
+    elder_worthy = True
+    for week in fame_hist:
+        if week is None or week < 1200:
```
Comment (damoster) [inline clash_royale_service.py:50]: Plz put this 1200 threshold in a const variable and put in schema.py also copy paste the comment explaining where this value comes from

File: clash_royale_service.py [hunk 22]
```diff
@@ -97,3 +129,47 @@ def get_clan_cards_rank(self, clan_tag, card_type_filter='all'):
...
+                    # Third Grab all the participants from the list and store it into the output object
+                    for participant in clan['participants']:
+                        if participant['tag'] in clan_players_war_history:
+                            clan_players_war_history[participant['tag']].fame_hist[current_week] = participant['fame']
+                            clan_players_war_history[participant['tag']].boat_attacks_hist[current_week] = participant['boatAttacks']
+                    break
+            current_week += 1
+        '''
+        Computing War active, elder worthy, and average fame
+        '''
```
Comment (damoster) [inline clash_royale_service.py:167]: /supernit does this comment really need 3 lines? Could just use # instead hahahaha

File: clash_royale_service.py (hunk 20 above)
Comment (josephwccheng) [reply clash_royale_service.py:42]: Done! MIN_FAME_WEEK

File: tests/test_clan_members_war.py [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+        when(clan_members_war.clash_royale_client).get_clan_info(clan_tag).thenReturn(
+            clash_royale_client_responses.CLAN_INFO_API_RESPONSE
+        )
+        when(clan_members_war.clash_royale_client).get_river_race_log(clan_tag).thenReturn(
+            clash_royale_client_riverracelog.RIVER_RACE_LOG_API_RESPONSE
+        )
+
+        # Testing if three players is returned based on get clan info
+        clan_players_war_history = clan_members_war.clan_river_race_history(clan_tag)
+        self.assertEqual(len(clan_players_war_history), 3)
```
Comment (damoster) [inline tests/test_clan_members_war.py:63]: For a sanity check, can you check just ONE player has the expected output e.g.
```python
joseph_activity = PlayerActivity(tag='#8VUG0GQRY', name='joseph', role='elder', exp_level=13, fame_hist=[2650, 2800, 2500, 3100], boat_attacks_hist=[0, 0, 0, 0], war_active=True, elder_worthy=True, avg_fame=2762.5)
self.assertEqual(clan_players_war_history['#8VUG0GQRY'], joseph_activity)
```

NOTE: no need repeat all the same edge cases as already tested in `test_past_weeks_clan_war`

Comment (damoster) [reply tests/test_clan_members_war.py:63]: Reason why is in `test_past_weeks_clan_war` you're mocking the response and passing into `past_weeks_clan_war`. Whereas in this test you're not, it's being computed from  `RIVER_RACE_LOG_API_RESPONSE`

File: common/schemas.py [hunk 24]
```diff
@@ -1,4 +1,32 @@
+from dataclasses import dataclass
+
 # Numbers represent the prefix of the id of the cards.
 # i.e. 26000037 is a Inferno Dragon, 27000007 is Elixir Collector, 28000000 is fireball
 CARD_TYPE_ID_PREFIX = {'26': 'troop', '27': 'building', '28': 'spell'}
 MAX_CARD_LEVEL = 14
+
+# Minimum fame week required for someone to be active. i.e. used 4 decks 3 times = 12 battles. 12 * 100 = 1200 fame.
```
Comment (damoster) [inline common/schemas.py:8]: /supernit maybe also indicate what 100 is (fame for losing non-boat battle)

**Review by damoster — APPROVED** [review]
Nice work. Merge away
